src/main/habit-modal-dialogue/habit-modal-dialogue.component.ts

```diff
@@ -0,0 +1,530 @@
+  }
+
+  updateWaterGoalFrequency(times: string) {
+    if (this.editModal) {
+      this.receivedHabit.Frequency = times;
+    } else {
+      this.waterFrequency = times;
+    }
+  }
+
+  isWaterOrRunningOrCycling(): boolean {
+    return this.habitName === AppConstants.drinkWater || this.habitName === AppConstants.running || this.habitName === AppConstants.cycling;
+  }
+
+  updateRunningGoalFrequency(value: string) {
+    if (this.editModal) {
+      this.receivedHabit.Frequency = value;
+    } else {
+      this.runningFrequency = value;
+    }
+  }
+
+  saveHabit() {
+
+    if (this.editModal) {
+      this.receivedHabit.name = this.habitName;
+      this.habitService.updateHabit(this.receivedHabit);
+
+    } else {
+      // Set goal if not provided
+      this.goal = this.goal || this.getHabitGoalValue();
+
+      // Set frequency if not provided
+      this.frequency = this.frequency || this.getHabitFrequency();
+
+      // Add habit to the service
+      this.habitService.addHabit(this.habitName, this.goal, this.frequency, this.frequencyPerPeriod, this.repeatValue, this.timeOfDay, this.startSelectedDate);
+    }
+    // Close the dialog
+    this.ref.close();
+  }
+
+  onGoalChange(value: number) {
+    setTimeout(() => {
+      this.updateGoal(value);
+    });
+  }
+
+  updateGoal(value: number) {
+
+    if (this.editModal) {
+      this.receivedHabit.goal = value;
+    } else {
+      this.goal = value;
+    }
+  }
+
+  updateFrequency(value: string) {
+
+    if (this.editModal) {
+      this.receivedHabit.Frequency = value;
+    } else {
+      this.frequency = value;
+    }
+  }
+
+  deleteHabit(habit: Habit) {
+
+    this.habitService.deleteHabit(habit);
+    this.ref.close();
+  }
+
+  archiveHabit() {
+    this.habitService.archiveHabit(this.receivedHabit);
+    this.ref.close();
+  }
+
+  isEditModal() {
+    return this.editModal;
+  }
+
+  checkIntervalPerDays(repeat: string) {
+    if (this.editModal) {
+
+      if (this.receivedHabit.repeat === repeat) {
+        return true;
+      }
+    } else {
+      if (this.intervalPerDays === repeat) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  onStartDateSelect(date: Date) {
+
+    if (this.editModal) {
+      this.receivedHabit.startDate = date;
+    } else {
+      this.selectedDate = date;
+    }
+  }
+
+  private getHabitGoalValue(): number {
+    switch (this.habitName) {
+      case AppConstants.drinkWater:
+        return this.goalWater;
+      case AppConstants.running:
+        return this.runningValue;
+      case AppConstants.cycling:
+        return this.cyclingValue;
+      default:
+        return this.goalValue;
+    }
+  }
+
+  private getHabitFrequency(): string {
+    switch (this.habitName) {
+      case AppConstants.drinkWater:
+        return this.waterFrequency;
+      case AppConstants.running:
+      case AppConstants.cycling:
+        return this.runningFrequency;
+      default:
+        return this.goalFrequency;
+    }
+  }
+}
```

Comment:
same as above as this component is very lengthy now and we should break it into multiple components